File: week5/FIBOSUM.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import sys
MOD = 1000000007
"""
Using matrix exponentiation, implement it with while loop rather than recursive function call
Refer to: http://fulmicoton.com/posts/fibonacci/   http://www.maths.surrey.ac.uk/hosted-sites/R.Knott/Fibonacci/fibFormula.html#exact 
http://nayuki.eigenstate.org/page/fast-fibonacci-algorithms
TIME: 4.4s MEM：4.0M 
STATUS:Accepted
DATE：April 8, 2014
From: http://www.spoj.com/submit/FIBOSUM/
"""
__author__ = 'frankfu'

#nonrecursive use while
def nonre_fibo_matrix(n):
	M = [[1,1], [1,0]] #or [[0,1], [1,1]] 
	ret = [[1,0], [0,1]]
	temp = [[0,0], [0,0]]
	while n:
		if n&1:
			temp = [[0,0], [0,0]]
			for i in range(2):
				for j in range(2):
					for k in range(2):
						temp[i][j] = (temp[i][j]+ret[i][k]*M[k][j])%MOD
			for i in range(2):
				for j in range(2):
					ret[i][j] = temp[i][j]
		n = n>>1
		if not n:
			break
		temp = [[0,0], [0,0]]
		for i in range(2):
			for j in range(2):
				for k in range(2):
					temp[i][j] = (temp[i][j]+M[i][k]*M[k][j])%MOD
		for i in range(2):
			for j in range(2):
				M[i][j] = temp[i][j]
	return (ret[0][1])%MOD
#t <= 1000
for t in range(int(sys.stdin.readline().strip())):
	#0<=N<=M<=10^9
	N,M = sys.stdin.readline().strip().split()
	N = int(N)
	M = int(M)
	#matrix exponentiation
	print (nonre_fibo_matrix(M+2)-nonre_fibo_matrix(N+1)+MOD)%MOD
	

File: week5/FIBOSUM.c
#include <stdio.h>
#define MOD 1000000007
#define LL long long 
/*
Using matrix exponentiation, implement it with while loop rather than recursive function call
Refer to: http://fulmicoton.com/posts/fibonacci/   http://www.maths.surrey.ac.uk/hosted-sites/R.Knott/Fibonacci/fibFormula.html#exact 
http://nayuki.eigenstate.org/page/fast-fibonacci-algorithms
TIME: 0.05s MEM：1.6M 
STATUS:Accepted
DATE：April 8, 2014
From: http://www.spoj.com/submit/FIBOSUM/
__author__ = 'frankfu'
*/
LL fib_matrix(int n)
{
	int i=0,j=0,k=0;
	LL fib[2][2]={{1,1},{1,0}}, ans[2][2]={{1,0},{0,1}}, temp[2][2]={{0,0},{0,0}};
	while(n)
	{
		if(n&1)
		{
			for(i=0;i<2;i++)
				for(j=0;j<2;j++)
					temp[i][j]=0;
			for(i=0;i<2;i++) 
				for(j=0;j<2;j++)
					for(k=0;k<2;k++)
						temp[i][j]=(temp[i][j]+ans[i][k]*fib[k][j])%MOD;
			for(i=0;i<2;i++) 
				for(j=0;j<2;j++) 
					ans[i][j]=temp[i][j];
		}
		n = n >> 1;
		if(!n) 
			break;
		for(i=0;i<2;i++)
			for(j=0;j<2;j++)
				temp[i][j]=0;
		for(i=0;i<2;i++) 
			for(j=0;j<2;j++) 
				for(k=0;k<2;k++)
					temp[i][j]=(temp[i][j]+fib[i][k]*fib[k][j])%MOD;
		for(i=0;i<2;i++)
			for(j=0;j<2;j++) 
				fib[i][j]=temp[i][j];
	}
	return (ans[0][1])%MOD;
}

int main()
{
	int t=0,M=0,N=0;
	scanf("%d", &t);
	while(t)
	{
		scanf("%d %d", &N, &M);
		printf("%lld\n", (fib_matrix(M+2)-fib_matrix(N+1)+MOD)%MOD);
		t--;
	}
	return 0;
}

